### PR TITLE
API work: refactors routers, adds controllers, expands db models

### PR DESCRIPTION
**Pull request by belcurv**
Architectural changes to the API:
* Rearranges some server logic (grouping imports at the top, middleware in the middle :wink:, etc)
* Removes unused router ("Hi Bears Team") we used to test way back in the beginning
* Refactors how routers are called. Express routers mount to paths, so I specify mount paths when initializing them. (ex: `app.use('/auth', require('./routes/authRoutes'))` )
* Refactors and/or adds routers in the router/controller/model style. Routers only call controller methods. Controllers call Model methods, etc.

New API stuff
* Adds a **Match** DB model
* Adds a `dbSeed` controller to seed our database with 100 match documents. These docs are generated randomly as either single or multi-player, with either `score` (single player) or `winner` properties (multi-player).
* Adds a **match** controller. It currently exports only a single method: `getLeaderboard`, which returns the top 10 single player scores and the top 10 most victorious multi-player users.
* Adds **profile** controller. It exports the old `getUserProfile` that was previously in-line in authRoutes.
* Adds DB setup util module. Why: so we can programmatically set `NODE_ENV` to 'development' or 'testing' to target different databases (local mongod vs Mlab).
* Adds `createdAt` and `updatedAt` timestamps to existing User DB model.
* Adds new route: `GET >> /api/leaderboard`. It calls `getLeaderboard` from the match controller.

The `GET /api/leaderboard` route returns JSON of this shape:

```
{
    "singlePlayer": [
        {
            "name": "Quincy",
            "score": 966
        },
        {
            "name": "Dereje",
            "score": 955
        },
        {
            "name": "terrance",
            "score": 938
        },
    ],
    "multiPlayer": [
        {
            "wins": 6,
            "name": "Dereje"
        },
        {
            "wins": 4,
            "name": "Quincy"
        },
        {
            "wins": 4,
            "name": "Kevin"
        },
    ]
}
```

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -25,6 +25,7 @@ module.exports = {
       { 'SwitchCase': 1 }
     ],
     'no-console': 0,
+    'no-underscore-dangle': ["error", { "allow": ["_id"] }],
```
Comment (terrancexin) [inline .eslintrc.js:28]: `id` does work for mongoose

Comment (belcurv) [reply .eslintrc.js:28]: Eslint complained for me. "no-underscore-dangle" may come from Airbnb or CRA linter rules. This rule overrides it just for `_id`

**Review by eolculnamo2 — APPROVED** [review]

**Comment by belcurv** [comment]
New route `GET >> /api/best_scores` - returns every player's best single-player score, sorted alphabetically by player name.

* Calls Match model static method `getAllBestSPScores()`.
* Takes no params/arguments/etc.
* Returns JSON in this shape:
```
[
    {
        "bestScore": 641,
        "_id": "5b4cd14d343984146b6dd67a",
        "name": "Adam"
    },
    {
        "bestScore": 865,
        "_id": "5b4cd14d343984146b6dd67d",
        "name": "Bernard"
    },
    {
        "bestScore": 746,
        "_id": "5b4cd14d343984146b6dd679",
        "name": "Chance"
    },
    {
        "bestScore": 703,
        "_id": "5b4cd14d343984146b6dd673",
        "name": "Danica"
    },
    ...
```

**Comment by belcurv** [comment]
I also refactored the match controller and model - moving all DB logic into Match schema static methods. This will make it easier to unit test our controllers, should we choose to add testing.

File: server/controllers/matchCtrl.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * Returns: JSON array of all players' best single-player scores
+*/
+const getPlayersBestScores = (req, res, next) => {
+  Match.getAllBestSPScores()
+    .then(bestScores => res.json(bestScores))
+    .catch(err => next(err));
+};
+
+/**
+ * Get all player's recent match results
```
Comment (belcurv) [inline server/controllers/matchCtrl.js]: Oops - this should read "Get one player's own recent match results"

Comment (belcurv) [reply server/controllers/matchCtrl.js]: Fixed in commit# 37efa75

**Pull request merged**